arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,501 @@
+
+  /* Load */
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rd     : 3;
+    uint16_t imm6   : 1; /* IMM bit [6] */
+    uint16_t imm2   : 1; /* IMM bit [2] */
+    uint16_t rs1    : 3;
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } lw;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t imm42  : 3; /* IMM bit [4:2] */
+    uint16_t rd     : 5;
+    uint16_t imm5   : 1; /* IMM bit [5] */
+    uint16_t funct3 : 3;
+  } lwsp;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rd     : 3;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t rs1    : 3;
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } ld;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t imm86  : 3; /* IMM bit [6:6] */
+    uint16_t imm43  : 2; /* IMM bit [4:3] */
+    uint16_t rd     : 5;
+    uint16_t imm5   : 1; /* IMM bit [5] */
+    uint16_t funct3 : 3;
+  } ldsp;
+
+  /* Store */
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 3;
+    uint16_t imm6   : 1; /* IMM bit [6] */
+    uint16_t imm2   : 1; /* IMM bit [2] */
+    uint16_t rs1    : 3;
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } sw;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 5;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t imm52  : 4; /* IMM bit [5:2] */
+    uint16_t funct3 : 3;
+  } swsp;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 3;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t rs1    : 3;
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } sd;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 5;
+    uint16_t imm86  : 3; /* IMM bit [8:6] */
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } sdsp;
+} riscv_insn_c_t;
+
+/* Normal instruction encoding */
+
+typedef union
+{
+  uint32_t insn;
+
+  /* Load */
+
+  struct
+  {
+    uint32_t op     : 7;
+    uint32_t rd     : 5;
+    uint32_t funct3 : 3;
+    uint32_t rs1    : 5;
+    uint32_t imm    : 12;
+  } l;
+
+  /* Store */
+
+  struct
+  {
+    uint32_t op     : 7;
+    uint32_t imm2   : 5;
+    uint32_t funct3 : 3;
+    uint32_t rs1    : 5;
+    uint32_t rs2    : 5;
+    uint32_t imm1   : 7;
+  } s;
+} rv_insn_t;
+
+typedef struct
+{
+  uint8_t *dest;
+  uint8_t *src;
+  int      len;
+  bool     sext;
+} riscv_insn_ctx_t;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sext
+ *
+ * Description:
+ *   Sign extension a given bit width data
+ *
+ ****************************************************************************/
+
+static intptr_t sext(intptr_t v, uint32_t w)
+{
+  w = 8 * sizeof(intptr_t) - w;
```

Comment:
how to handle 64bit sign extension on 32bit?

Reply:
RV32 don't have 64bit load instruction, so we don't need do 64 bit sext on 32 bit platform